arbos/multigasCollector/collector_test.go

```diff
@@ -0,0 +1,522 @@
+						BlockTimestamp: 1234567890,
+					},
+				},
+				{Type: CollectorMsgStartBlock},
+				{
+					Type: CollectorMsgTransaction,
+					Transaction: &TransactionMultiGas{
+						TxHash:  []byte{0x78, 0x9a, 0xbc},
+						TxIndex: 1,
+						MultiGas: *multigas.MultiGasFromMap(map[multigas.ResourceKind]uint64{
+							multigas.ResourceKindComputation:   200,
+							multigas.ResourceKindHistoryGrowth: 60,
+							multigas.ResourceKindStorageAccess: 120,
+						}),
+					},
+				},
+				{
+					Type: CollectorMsgFinaliseBlock,
+					Block: &BlockInfo{
+						BlockNumber:    12346,
+						BlockHash:      []byte{0xde, 0xf1, 0x23},
+						BlockTimestamp: 1234567891,
+					},
+				},
+				{Type: CollectorMsgStartBlock},
+				{
+					Type: CollectorMsgTransaction,
+					Transaction: &TransactionMultiGas{
+						TxHash:  []byte{0xab, 0xcd, 0xef},
+						TxIndex: 2,
+						MultiGas: *multigas.MultiGasFromMap(map[multigas.ResourceKind]uint64{
+							multigas.ResourceKindComputation:   300,
+							multigas.ResourceKindStorageGrowth: 800,
+							multigas.ResourceKindUnknown:       35,
+						}),
+					},
+				},
+				{
+					Type: CollectorMsgFinaliseBlock,
+					Block: &BlockInfo{
+						BlockNumber:    12347,
+						BlockHash:      []byte{0x45, 0x67, 0x89},
+						BlockTimestamp: 1234567892,
+					},
+				},
+			},
+			expectFiles: 2,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			tmpDir := t.TempDir()
+			input := make(chan *CollectorMessage, 10)
+
+			config := Config{
+				OutputDir: tmpDir,
+				BatchSize: int(tt.batchSize), //nolint:gosec
+			}
+
+			c, err := NewCollector(config, input)
+			require.NoError(t, err)
+			c.Start(context.Background())
+
+			for _, msg := range tt.inputData {
+				input <- msg
+			}
+			close(input)
+			c.StopAndWait()
+
+			files, err := filepath.Glob(filepath.Join(tmpDir, "multigas_batch_*.pb"))
+			require.NoError(t, err)
+			assert.Len(t, files, tt.expectFiles)
+
+			// Decode batches from files
+			var allData []*proto.BlockMultiGasData
+			for _, file := range files {
+				data, err := os.ReadFile(file)
+				require.NoError(t, err)
+				var batch proto.BlockMultiGasBatch
+				require.NoError(t, protobuf.Unmarshal(data, &batch))
+				allData = append(allData, batch.Data...)
+			}
+
+			// Build expected blocks: group TXs until FinaliseBlock; reset on StartBlock
+			type expBlock struct {
+				block *BlockInfo
+				txs   []*TransactionMultiGas
+			}
+			var expected []expBlock
+			var curTxs []*TransactionMultiGas
+			for _, m := range tt.inputData {
+				switch m.Type {
+				case CollectorMsgStartBlock:
+					// Explicitly drop any unfinalised txs (mirrors collector behavior)
+					curTxs = nil
+				case CollectorMsgTransaction:
+					curTxs = append(curTxs, m.Transaction)
+				case CollectorMsgFinaliseBlock:
+					expected = append(expected, expBlock{block: m.Block, txs: curTxs})
+					curTxs = nil
+				}
+			}
+
+			assert.Len(t, allData, len(expected))
```

Comment:
I really think there is a protocmp package that knows how to compare protocol buffers for equality. It should greatly simplify this test. Let me see if I can find it. https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp